readalongs/align.py

```diff
@@ -33,46 +33,58 @@
 from readalongs.text.util import save_xml
 from readalongs.log import LOGGER
 
-####
-#
-# Some distros (Python2, Python3 on Windows it seems) don't have the WAV
-# reading methods being context managers; the following checks whether the
-# necessary methods are present and if not, add thems.
-#
-# Based on http://web.mit.edu/jgross/Public/21M.065/sound.py 9-24-2017
-####
-
 
-def _trivial__enter__(self):
-    return self
+def correct_adjustments(start: Union[int, float], end: Union[int, float], do_not_align_segments: List[object]) -> (int, int):
+    """ Given the start and end of a segment and a list of do-not-align segments,
+        If one of the do-not-align segments occurs inside one of the start-end range,
+        align the start or end with the do-not-align segment, whichever requires minimal change
+    """
+    if isinstance(start, float):
+        LOGGER.info(f'Converting {start} to milliseconds')
+        start = int(start * 1000)
+    if isinstance(end, float):
+        LOGGER.info(f'Converting {end} to milliseconds')
+        end = int(end * 1000)
+    for seg in do_not_align_segments:
+        if start < seg['begin'] and end > seg['end']:
+            if seg['begin'] - start > end - seg['end']:
+                return start, seg['begin']
+            else:
+                return seg['end'], end
+    return start, end
 
-def _self_close__exit__(self, exc_type, exc_value, traceback):
-    self.close()
 
-if not hasattr(wave.Wave_read, "__exit__"):
-    wave.Wave_read.__exit__ = _self_close__exit__
-if not hasattr(wave.Wave_write, "__exit__"):
-    wave.Wave_write.__exit__ = _self_close__exit__
-if not hasattr(wave.Wave_read, "__enter__"):
-    wave.Wave_read.__enter__ = _trivial__enter__
-if not hasattr(wave.Wave_write, "__enter__"):
-    wave.Wave_write.__enter__ = _trivial__enter__
+def calculate_adjustment(timestamp: Union[int, float], do_not_align_segments: List[object]) -> int:
+    """ Given a time and a list of do-not-align segments,
+        return the sum (ms) of the lengths of the do-not-align segments 
+        that start before the timestamp
+    """
+    if isinstance(timestamp, float):
+        LOGGER.info(f'Converting {timestamp} to milliseconds')
+        timestamp = int(timestamp * 1000)
+    return sum(
+        (seg['end'] - seg['begin'])
+        for seg in do_not_align_segments
+        if seg['begin'] <= timestamp
+    )
 
 
-def align_audio(xml_path: str, wav_path: str, unit:str ='w', bare=False, save_temps:Union[str, None] = None):
+def align_audio(xml_path: str, audio_path: str, unit: str = 'w', bare=False, config=None, save_temps: Union[str, None] = None):
     """ Align an XML input file to an audio file.
-    
+
     Parameters
     ----------
     xml_path : str
         Path to XML input file in TEI-like format
-    wav_path : str
-        Path to audio input (WAV or MP3)
+    audio_path : str
+        Path to audio input
```

Comment:
It might still be useful to put an indication here as to which audio file formats are supported. Maybe "any format supported by X" where X is the library we're using?

Reply:
Good idea! We're using pydub which I believe just uses the system installation of ffmpeg, so we can put that there.